### DISCUSSION TITLE
tutorials/lift-off-part4/03-adding-a-mutation-to-our-schema

### DISCUSSION THREAD

**Discussion by giscus** [apollographql-education/content-feedback#78] [course feedback]
# tutorials/lift-off-part4/03-adding-a-mutation-to-our-schema

https://www.apollographql.com/tutorials/lift-off-part4/03-adding-a-mutation-to-our-schema

**Comment by RumualdoPablo** [comment]
Great content!

**Comment by belloshehu** [comment]
Awesome!

**Comment by omriluz** [comment]
Thanks for the lesson :)

**Comment by torressam333** [comment]
At what point do we separate out Query and Mutations into separate files so that our schema.js/ts doesn't become gigantic?

**Reply by mabuyo** [reply]
Great question! It depends on the project and your team. A one schema file setup works fine for smaller projects and solo devs, but as the API grows, that might become harder to manage, scale, deploy. For larger projects or multiple teams, we recommend using a [federated GraphQL architecture](https://www.apollographql.com/docs/federation). In this case, we're splitting the schema by business domains and team ownership (not types, like Query and Mutation). 

Hope that helps, feel free to check out the [Voyage I](https://www.apollographql.com/tutorials/voyage-part1) course. If you're curious to see just a schema split up into different files to avoid an overly-long file, there's a [blog post that outlines the process](https://www.apollographql.com/blog/modularizing-your-graphql-schema-code).

**Comment by Chip-L** [comment]
Is there a way to extend a type? If we decided we always want to return `code`, `success`, and `message`, it seems like we should be able to create a `MutationResponse` type and then just extend it. Then we could have something like:

```
  type MutationResponse {
    code: Int!
    success: Boolean!
    message: String!
  };

  type IncrementTrackViewsResponse extends MutationResponse {
     track: Track
  }

  type AssignSpaceshipResponse extends MutationResponse {
    spaceship: Spaceship
    mission: Mission
  }
```

**Reply by mabuyo** [reply]
Hey there @Chip-L! I think what you're looking for is the `interface` type. That's definitely another way we could define `MutationResponse`, if we wanted to make sure that every single return type for a mutation included those 3 fields.

We'd still need to explicitly define the `code`, `success` and `message` fields in the implementing types. If we didn't, GraphQL throws an error to force us to do so.

```graphql
  interface MutationResponse {
    code: Int!
    success: Boolean!
    message: String!
  };

  type IncrementTrackViewsResponse implements MutationResponse {
    code: Int!
    success: Boolean!
    message: String!
     track: Track
  }
```

**Comment by gustvoassuncao** [comment]
Thanks for the lesson! Awesome as always!

**Comment by HuyAms** [comment]
Hi, on error, why do we decide to return a response data with `success: false` instead of throwing a GraphQLError?

**Reply by lizhennessy** [reply]
Hi there @HuyAms ! This is a great question. There are lots of ways we can handle errors, and it depends on what you want your end user experience to look like. With this approach, we define a standard response shape that we can expect from running a mutation - whether it succeeds or fails. This lets us handle the error within the resolver, even determine how much of the error message we _want_ to show, or even just hide all those details with a simple `"success: false"`. This can be a helpful assurance for your downstream consumers that the response from running a mutation will always have a particular shape they can rely on. You can of course choose to throw an error instead if that's the approach that works better for you and consumers of your API! Last week at GraphQL Summit there was [a great talk on exactly this subject](https://reg.rainfocus.com/flow/apollo/summit24/graphqlsummitvirtual24/page/virtualcatalog/session/1726847795675001byJ0); there's going to be [a virtual summit in the next month](https://summit.graphql.com/virtual), so I'd encourage you to register and attend if you're interested!
